File: web-server/test/testutil/minio_helper.go
package testutil

import (
	"context"
	"io"
	"os"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

// MinIOHelper provides direct MinIO access for test verification
type MinIOHelper struct {
	client     *minio.Client
	bucketName string
}

// NewMinIOHelper creates a new MinIO helper for tests
func NewMinIOHelper() (*MinIOHelper, error) {
	endpoint := os.Getenv("NK_MINIO_ENDPOINT")
	if endpoint == "" {
		endpoint = "localhost:9000"
	}

	accessKey := os.Getenv("NK_MINIO_ACCESS_KEY")
	if accessKey == "" {
		accessKey = "minioadmin"
	}

	secretKey := os.Getenv("NK_MINIO_SECRET_KEY")
	if secretKey == "" {
		secretKey = "minioadmin"
	}

	bucketName := os.Getenv("NK_MINIO_BUCKET_NAME")
	if bucketName == "" {
		bucketName = "nkrypt-xyz-dev"
	}

	useSSL := os.Getenv("NK_MINIO_USE_SSL") == "true"

	client, err := minio.New(endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(accessKey, secretKey, ""),
		Secure: useSSL,
	})
	if err != nil {
		return nil, err
	}

	return &MinIOHelper{
		client:     client,
		bucketName: bucketName,
	}, nil
}

// BlobExists checks if a blob exists in MinIO
func (m *MinIOHelper) BlobExists(ctx context.Context, blobID string) (bool, error) {
	objectName := "blobs/" + blobID
	_, err := m.client.StatObject(ctx, m.bucketName, objectName, minio.StatObjectOptions{})
	if err != nil {
		errResp := minio.ToErrorResponse(err)
		if errResp.Code == "NoSuchKey" || errResp.Code == "NoSuchBucket" {
			return false, nil
		}
		return false, err
	}
	return true, nil
}

// GetBlobSize returns the size of a blob in MinIO
func (m *MinIOHelper) GetBlobSize(ctx context.Context, blobID string) (int64, error) {
	objectName := "blobs/" + blobID
	info, err := m.client.StatObject(ctx, m.bucketName, objectName, minio.StatObjectOptions{})
	if err != nil {
		errResp := minio.ToErrorResponse(err)
		if errResp.Code == "NoSuchBucket" {
			return 0, nil
		}
		return 0, err
	}
	return info.Size, nil
}

// GetBlob retrieves blob data from MinIO
func (m *MinIOHelper) GetBlob(ctx context.Context, blobID string) ([]byte, error) {
	objectName := "blobs/" + blobID
	obj, err := m.client.GetObject(ctx, m.bucketName, objectName, minio.GetObjectOptions{})
	if err != nil {
		return nil, err
	}
	defer obj.Close()

	return io.ReadAll(obj)
}

// DeleteBlob removes a blob from MinIO (for cleanup)
func (m *MinIOHelper) DeleteBlob(ctx context.Context, blobID string) error {
	objectName := "blobs/" + blobID
	return m.client.RemoveObject(ctx, m.bucketName, objectName, minio.RemoveObjectOptions{})
}

// ListBlobs lists all blobs with a given prefix
func (m *MinIOHelper) ListBlobs(ctx context.Context, prefix string) ([]string, error) {
	var blobs []string
	objectCh := m.client.ListObjects(ctx, m.bucketName, minio.ListObjectsOptions{
		Prefix:    "blobs/" + prefix,
		Recursive: true,
	})

	for object := range objectCh {
		if object.Err != nil {
			return nil, object.Err
		}
		// Remove "blobs/" prefix from the key
		blobID := object.Key[6:]
		blobs = append(blobs, blobID)
	}

	return blobs, nil
}

File: independent-integration-test/test/testlib/minio-helper.js
import * as Minio from 'minio';
import { Readable } from 'stream';

/**
 * MinIO helper for test verification
 * This allows tests to verify that blobs are correctly stored in MinIO
 */
export class MinIOHelper {
  constructor(config) {
    this.client = new Minio.Client({
      endPoint: config.endpoint || 'localhost',
      port: config.port || 9000,
      useSSL: config.useSSL || false,
      accessKey: config.accessKey || 'minioadmin',
      secretKey: config.secretKey || 'minioadmin',
    });
    this.bucketName = config.bucketName || 'nkrypt-blobs';
  }

  /**
   * Ensure the bucket exists
   */
  async ensureBucket() {
    const exists = await this.client.bucketExists(this.bucketName);
    if (!exists) {
      await this.client.makeBucket(this.bucketName);
    }
  }

  /**
   * Get a blob from MinIO as a Buffer
   * @param {string} blobId - The blob ID
   * @returns {Promise<Buffer>} The blob data
   */
  async getBlob(blobId) {
    const objectKey = `blobs/${blobId}`;
    
    return new Promise((resolve, reject) => {
      const chunks = [];
      
      this.client.getObject(this.bucketName, objectKey, (err, stream) => {
        if (err) {
          return reject(err);
        }

        stream.on('data', (chunk) => chunks.push(chunk));
        stream.on('end', () => resolve(Buffer.concat(chunks)));
        stream.on('error', reject);
      });
    });
  }

  /**
   * Get a blob from MinIO as a readable stream
   * @param {string} blobId - The blob ID
   * @returns {Promise<Readable>} The blob stream
   */
  async getBlobStream(blobId) {
    const objectKey = `blobs/${blobId}`;
    
    return new Promise((resolve, reject) => {
      this.client.getObject(this.bucketName, objectKey, (err, stream) => {
        if (err) {
          return reject(err);
        }
        resolve(stream);
      });
    });
  }

  /**
   * Get blob metadata (size, etc.)
   * @param {string} blobId - The blob ID
   * @returns {Promise<Object>} The blob stat info
   */
  async getBlobStat(blobId) {
    const objectKey = `blobs/${blobId}`;
    return await this.client.statObject(this.bucketName, objectKey);
  }

  /**
   * Check if a blob exists
   * @param {string} blobId - The blob ID
   * @returns {Promise<boolean>} True if blob exists
   */
  async blobExists(blobId) {
    try {
      await this.getBlobStat(blobId);
      return true;
    } catch (err) {
      if (err.code === 'NotFound') {
        return false;
      }
      throw err;
    }
  }

  /**
   * Delete a blob from MinIO
   * @param {string} blobId - The blob ID
   */
  async deleteBlob(blobId) {
    const objectKey = `blobs/${blobId}`;
    await this.client.removeObject(this.bucketName, objectKey);
  }

  /**
   * List all blobs with a given prefix
   * @param {string} prefix - The prefix to filter by (default: 'blobs/')
   * @returns {Promise<Array>} Array of object names
   */
  async listBlobs(prefix = 'blobs/') {
    return new Promise((resolve, reject) => {
      const objects = [];
      const stream = this.client.listObjects(this.bucketName, prefix, true);
      
      stream.on('data', (obj) => objects.push(obj));
      stream.on('end', () => resolve(objects));
      stream.on('error', reject);
    });
  }

  /**
   * Clean up all test blobs
   */
  async cleanupAllBlobs() {
    const objects = await this.listBlobs('blobs/');
    const objectNames = objects.map(obj => obj.name);
    
    if (objectNames.length > 0) {
      await this.client.removeObjects(this.bucketName, objectNames);
    }
  }
}

/**
 * Create a MinIO helper with default test configuration
 * Can be overridden with environment variables:
 * - MINIO_ENDPOINT
 * - MINIO_PORT
 * - MINIO_ACCESS_KEY
 * - MINIO_SECRET_KEY
 * - MINIO_BUCKET_NAME
 */
export function createTestMinIOHelper() {
  return new MinIOHelper({
    endpoint: process.env.MINIO_ENDPOINT || 'localhost',
    port: parseInt(process.env.MINIO_PORT || '9000'),
    useSSL: process.env.MINIO_USE_SSL === 'true',
    accessKey: process.env.MINIO_ACCESS_KEY || 'minioadmin',
    secretKey: process.env.MINIO_SECRET_KEY || 'minioadmin',
    bucketName: process.env.MINIO_BUCKET_NAME || 'nkrypt-blobs',
  });
}
